### ISSUE TITLE
Pods with zero terminationGracePeriod are force-deleted

### ISSUE THREAD

**Issue by jsafrane** [kubernetes/kubernetes#120671]
### What happened?

This is spin-off from https://github.com/kubernetes/kubernetes/issues/119273.

When a pod with `spec.terminationGracePeriod: 0` is deleted without force, it's force-deleted from the API server, without kubelet killing its containers and unmounting its volumes first.

This can be dangerous for StatefulSets. They should guarantee that only a single replica of a Pod runs, but KCM will start a new replica while the old replica is still running, because kubelet did not have enough time to kill it and unmount its volumes. This can be further emphasized when kubelet is not able to talk to the API server e.g. due to network partition or any other error. The old Pod can run for quite some time until its kubelet reconnects to the API server and realizes it should be killed, while the new Pod is running on another node. These two pods are sharing the same volume and may corrupt data of each other.

Surprisingly, `terminationGracePeriod: 1` is enough to prevent this behavior. The old Pod will be Terminating without any timeout (i.e. forver) until kubelet confirms the Pod is dead and its volumes unmounted.

### What did you expect to happen?

Pods should be deleted from the API server only after their containers are killed and their volumes unmounted.

Unless `--force` is used, but then user must deal with the consequences - pod volumes can get corrupted, as they may be used by multiple pods.

### How can we reproduce it (as minimally and precisely as possible)?

1. Run a StatefulSet with a single replica with `terminationGracePeriod: 0`.
2. Stop kubelet on the node where the replica starts .
3. After 5-6 minutes, node controller deletes the pod (without force), but the pod gets deleted from the API server because of `terminationGracePeriod: 0`.
4. Observe that a new replica starts, while the old replica is still running on the "problematic" node.

Check the same with `terminationGracePeriod: 1` - the old pod is *not* force-deleted and StatefulSet ensures that only a single replica runs. The Pod will be Terminating forever, or until the node reconnects to the cluster and kubelet kills the pod containers + unmounts its volumes.

### Kubernetes version

<details>

```console
$ kubectl version
Kubernetes Version: v1.27.4+2c287eb
```

</details>

### Cloud provider

<details>
Not relevant, but it was AWS.
</details>

### OS version

<details>
PRETTY_NAME="Red Hat Enterprise Linux CoreOS 414.92.202309121329-0 (Plow)"
</details>

### Install tools

<details>
OpenShift
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by jsafrane** [comment]
This behavior is controller by Pod's strategy: https://github.com/kubernetes/kubernetes/blob/4abf29c5c86349c7d2c769d267f4552ced5d0c15/pkg/registry/core/pod/strategy.go#L174-L176

Not sure what SIG is responsible for it. Starting with:
/sig api-machinery
/sig architecture

**Comment by jsafrane** [comment]
/remove-sig node
This is not related to node in any way. Pods are deleted in the API server and as I pointed above and node / kubelet cannot do anything about it. Kubelet can be even unavailable at that point.

**Comment by gjkim42** [comment]
FYI:
https://github.com/kubernetes/kubernetes/blob/f7dacb63cc0b0e79450b2f6d97b1c6e6c46070bd/staging/src/k8s.io/api/core/v1/types.go#L3354-L3363

---

I am not against this but I think this is not a simple change.

Force deleting the pod with 0s of terminationGracePeriod and gracefully deleting the pod with 1s of terminationGracePeriod is a completely different operation.
If the pod cannot be cleaned up completely from its node(kubelet) for any reason, the latter one may be stuck forever.
(e.g. if the node does not respond...)

Changing this behavior will surprise many users.
(including our e2e tests as well)
https://github.com/kubernetes/kubernetes/blob/f7dacb63cc0b0e79450b2f6d97b1c6e6c46070bd/test/e2e/framework/pod/resource.go#L291-L303

cc @liggitt

**Comment by liggitt** [comment]
`terminationGracePeriod: 0` is the pod author making pods default to force deletion. It's a bad idea (for the reasons you outline in the issue), but we've let them express that, so we need a really really strong case for not letting them express that any more.

We could start by issuing warnings whenever an object with `terminationGracePeriod: 0` is created (or updated to change to `terminationGracePeriod: 0`).

cc @smarterclayton @bobbypage

**Comment by liggitt** [comment]
/remove-sig api-machinery
/sig node

Node owns the pod API surface

**Comment by jsafrane** [comment]
> terminationGracePeriod: 0 is the pod author making pods default to force deletion

I disagree here. API says it means that pod's containers will be killed without waiting. That's IMO very different from force-deleting pods from the API server.

https://github.com/kubernetes/kubernetes/blob/97e87e2c40e5b83399a44738d38653fd59c58e99/staging/src/k8s.io/api/core/v1/types.go#L3355-L3356

> Force deleting the pod with 0s of terminationGracePeriod and gracefully deleting the pod with 1s of terminationGracePeriod is a completely different operation.
If the pod cannot be cleaned up completely from its node(kubelet) for any reason, the latter one may be stuck forever.
(e.g. if the node does not respond...)

If a Pod gets stuck Terminating with `terminationGracePeriod: 0`, i.e. at unmounting its volumes, user still has option to force-delete the pod with `medatata.deletionGracePeriodSeconds: 0`.

**Comment by jsafrane** [comment]
> TerminationGracePeriodSeconds defines the default value for the delete operation's GracePeriodSeconds

If that's true, then it should be explicitly written in the field description, with a huge warning about zero value. Now it confuses people with container kill timeout.

Also, with `TerminationGracePeriodSeconds: 1` I would then expect that the pod is force-deleted from the API server after 1 second, regardless of what happens in kubelet. It's not deleted, it waits for volumes to unmount, possibly indefinitely (which is IMO right, but it seems I am minority here).

> If something in the kubelet is not actually terminating pods that are force-deleted from the API server, or is leaving hung volumes not represented in the API, that's a bug we should fix.

The worst that can happen is that kubelet cannot talk to the API server at that particular moment and it does not know its pod was force-deleted. So the pod keeps running, with its volumes still mounted, while it has been deleted from the API server. If it was part of a StatefulSet, KCM will run a new replica, breaking StatefulSet guarantees that replicas are unique and corrupting volumes.

**Comment by liggitt** [comment]
> > TerminationGracePeriodSeconds defines the default value for the delete operation's GracePeriodSeconds
> 
> If that's true, then it should be explicitly written in the field description, with a huge warning about zero value. Now it confuses people with container kill timeout.

+100 to documenting _and_ emitting API warnings when setting `TerminationGracePeriodSeconds: 0`, which we can do without *any* compatibility concerns.

**Comment by liggitt** [comment]
we could *maybe* consider some additional validation specifically around statefulsets (or translation of a <= 0 terminationGracePeriodSeconds value in their pod spec to a field to a `terminationGracePeriodSeconds: 1` value in the pods the StatefulSet controller creates) since they have specific guarantees about one-of behavior for their replicas, but we have to be *really* careful not to do things that would require end-users to create their own cleanup flows to avoid getting stuck

**Comment by pohly** [comment]
/reopen

DRA is also affected by this because resource tracking depends on not removing pods prematurely before the node is done with them.

I got bitten by this three years ago, and again this week (https://github.com/kubernetes/kubernetes/issues/131513).

**Comment by pohly** [comment]
FWIW, both better docs and the warning would have helped me: I was looking up the documentation to find the exact field name and then used kubectl where the warning would have been visible.

👍 for doing both. I think @bart0sh wants to look into it.

**Comment by SergeyKanzhelev** [comment]
/remove-lifecycle stale

@bart0sh are you still interested to look into better explaining this behavior to end user? Perhaps we can add an event to this Pod from the kubelet saying that Pod was force-deleted and may have leaked resources.
